tracing-subscriber/src/registry/fmt.rs

```diff
@@ -0,0 +1,349 @@
+            let fmt_fields = FormattedFields {
+                fmt_fields: buf,
+                _format_event: PhantomData::<fn(N)>,
+            };
+            extensions.insert(fmt_fields);
+        }
+    }
+
+    fn on_record(&self, id: &Id, values: &Record<'_>, ctx: Context<'_, S>) {
+        let span = ctx.span(id).expect("Span not found, this is a bug");
+        let mut extensions = span.extensions_mut();
+
+        let mut buf = String::new();
+        if self.fmt_fields.format_fields(&mut buf, values).is_ok() {
+            let buf = match extensions.get_mut::<FormattedFields<Self>>() {
+                Some(fields) => format!("{}{}", fields.fmt_fields, buf),
+                None => buf,
+            };
+            let fmt_fields = FormattedFields {
+                fmt_fields: buf,
+                _format_event: PhantomData::<fn(N)>,
+            };
+            extensions.insert(fmt_fields);
+        }
+    }
+
+    fn on_event(&self, event: &Event<'_>, ctx: Context<'_, S>) {
+        thread_local! {
+            static BUF: RefCell<String> = RefCell::new(String::new());
+        }
+
+        BUF.with(|buf| {
+            let borrow = buf.try_borrow_mut();
+            let mut a;
+            let mut b;
+            let mut buf = match borrow {
+                Ok(buf) => {
+                    a = buf;
+                    &mut *a
+                }
+                _ => {
+                    b = String::new();
+                    &mut b
+                }
+            };
+
+            if (self.is_interested)(event) {
+                let ctx = self.make_ctx(ctx);
+                if self.fmt_event.format_event(&ctx, &mut buf, event).is_ok() {
+                    let mut writer = self.make_writer.make_writer();
+                    let _ = io::Write::write_all(&mut writer, buf.as_bytes());
+                }
+            }
+
+            buf.clear();
+        });
+    }
+}
+
+/// `FmtContext` is used to propogate subscriber context to tracing_subscriber::fmt.
+pub struct FmtContext<'a, S, N> {
+    pub(crate) ctx: Context<'a, S>,
+    pub(crate) fmt_fields: &'a N,
+}
+
+impl<'a, S, N> fmt::Debug for FmtContext<'a, S, N> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("FmtContext").finish()
+    }
+}
+
+impl<'a, S, N> FormatFields<'a> for FmtContext<'a, S, N>
+where
+    S: Subscriber + for<'lookup> LookupSpan<'lookup> + LookupMetadata,
+    N: for<'writer> FormatFields<'writer> + 'static,
+{
+    fn format_fields<R: RecordFields>(
+        &self,
+        writer: &'a mut dyn fmt::Write,
+        fields: R,
+    ) -> fmt::Result {
+        self.fmt_fields.format_fields(writer, fields)
+    }
+}
+
+impl<'a, S, N> FmtContext<'a, S, N>
+where
+    S: Subscriber + for<'lookup> LookupSpan<'lookup> + LookupMetadata,
+    N: for<'writer> FormatFields<'writer> + 'static,
+{
+    /// Visits parent spans. Used to visit parent spans when formatting spans
```

Comment:
Nit: can there be a newline between "Visits parent spans." and the next sentence?

Reply:
also, this visits the current span _followed_ by all its parents, so I don't think "Visits parent spans" is actually correct...